CEL expressions approver design

erikgb opened the pull request:
This PR contains an initial design proposal allowing policy authors to use CEL expressions to express policy rules. If this is ever implemented, it should address https://github.com/cert-manager/approver-policy/issues/62.

Also includes an initial design template copied from [cert-manager](https://github.com/cert-manager/cert-manager/blob/master/design/template.md).

/cc @inteon @SgtCoDFish

erikgb commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,368 @@
+spec:
+  allowed:
+    dnsNames:
+      expressions:
+        - "self.endsWith(cr.namespace + '.apps.my-cluster.com')"
```
```suggestion
      validations:
        - rule: "self.endsWith(cr.namespace + '.apps.my-cluster.com')"
          message: "DNSName should be in cluster apps domain"
```

After a bit of thinking, I think we should align our design closer to upstream [validation rules](https://kubernetes.io/docs/tasks/extend-kubernetes/custom-resources/custom-resource-definitions/#validation-rules).

I'll update the suggested design now.

erikgb replied:
Modified the proposal to reuse the Kubernetes CEL validations.

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,368 @@
+To configure valid CSR attribute values, the API provides fields for allowed
+`values` (array; any element matching the attribute is sufficient),
+`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the request attribute.
+Allowed string fields accept wildcards "*" within its value(s).
```
```suggestion
Allowed string fields accept wildcards `"*"` within its value(s).
```
? Or `*` ...

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,368 @@
+While approver-policy allows policies to include namespace selectors, this does not scale when attempting to
+enforce CSR attributes that must be a function of the CertificateRequest namespace (or other CertificateRequest fields).
+This proposal suggests an extension to the CertificateRequestPolicy API allowing policy authors to specify approval rules using
+[CEL expressions](https://github.com/google/cel-spec) containing variables with information from the CertificateRequest.
```
Is there a reason to point to that instead of https://kubernetes.io/docs/reference/using-api/cel/?

erikgb replied:
We could link to both. The Kubernetes docs give a good overview, but if you're interested in the details, I think you need to consult the detailed spec.

erikgb replied:
Good point, thanks! I will change the link in my next round on this. Also thanks for the rewording suggestions! 👍

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,368 @@
+
+As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
+self-provisioning of server certificates. Since the cluster is multi-tenant, I want to ensure that tenants do not
+"tip on each other's toes" and enforce that the issuer only signs certificates where the X.509 DNS SAN attribute
```
```suggestion
"trip on each other's toes" and enforce that the issuer only signs certificates where the X.509 DNS SAN attribute
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,368 @@
+#### SPIFFE ID identifying namespace
+
+As a cluster admin in a multi-tenant cluster, I want to provide a cert-manager ClusterIssuer to our users allowing
+self-provisioning of X.509 SVID certificates. The SPIFFE ID (X.509 URI SAN) should identify the namespace
```
It'd be nice if `SVID` and `SPIFFE ID` were links to things.

erikgb replied:
Thanks @jsoref! Updated with links.

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,369 @@
+-->
+
+An [experimental approver-policy plugin](https://github.com/erikgb/cel-approver-policy-plugin) providing some of the
+requested features exists, and it's proposed to base the design on this plugin.
```
this line is awkward. https://en.wikipedia.org/wiki/Garden-path_sentence

In general, I'd discourage the use of abbreviations (`'`), 
```suggestion
requested features exists, and we propose to base the design on this plugin.
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,369 @@
+
+- More complex expressions that will require more in-depth knowledge of CEL.
+- Will create a "split" API, since you would probably use either `expressions` or `allowed`/`constraints`.
+  This could increase the maintenance burden and source to subtle bugs. Breaking the API could be an option to mend this.
```
```suggestion
  This could increase the maintenance burden and lead to subtle bugs. Breaking the API could be an option to mitigate this.
```

erikgb commented:
I have now updated the proposal and moved the previous proposal into the alternatives. The new proposal is now significantly closer to the upstream CEL validations rules API. Please take another look. I would love to see some progress on this design proposal so we eventually can move over to implementing it. 😸 

/cc @jsoref @inteon @SgtCoDFish @wallrj

SgtCoDFish reviewed (commented):
/lgtm
/approve
/hold

(Holding so others can chime in)

First of all, this is well-written and makes sense to me. I understand the rationale and thank you for your patience with this!

I think the idea here is solid. It adds a lot of flexibility and power and the tradeoff in complexity is worth it.

A big part of why I'm really relaxed about this is that approver-policy isn't super-widely used and we're still at v1alpha1, which to me means we should be more relaxed about making changes and trying things out.

I think we should proceed with this on the basis that we _should_ be experimenting with this. I'd also support making these changes in v1alpha2, leaving v1alpha1 as what we have today, which in my mind should make this even easier to approve. But I don't think that's required - I'd be happy to add this to v1alpha1.

As for the alternatives: I instinctively prefer having expressions be tied to fields rather than operating on the whole CSR (i.e. I prerfer the proposal to that alternative). I think the simpler the CEL the better.

inteon reviewed (approved):
+1

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+This template is adapted from Kubernetes Enchancements KEP template https://raw.githubusercontent.com/kubernetes/enhancements/a86942e8ba802d0035ec7d4a9c992f03bca7dce9/keps/NNNN-kep-template/README.md
+-->
+
+# CEL validation rules to approve or deny CertificateRequest
```
```suggestion
# CEL validation rules to approve or deny a CertificateRequest
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+-->
+
+While approver-policy allows policies to include namespace selectors, this does not scale when attempting to
+enforce CSR attributes that must be a function of the CertificateRequest namespace (or other CertificateRequest fields).
```
```suggestion
enforce CSR attributes that must be a function of the CertificateRequest's namespace (or other CertificateRequest fields).
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+know that this has succeeded?
+-->
+
+- allow policy authors to express `allowed` CSR attribute value(s) to be a function of CertificateRequest fields using CEL expressions
```
```suggestion
- allow policy authors to express `allowed` CSR attribute value(s) to be a function of CertificateRequest's fields using CEL expressions
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+CertificateRequest, and will deny the request if it contains an attribute which is not present in the `allowed` block.
+To configure valid CSR attribute values, the API provides fields for allowed
+`values` (array; any element matching the attribute is sufficient),
+`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the request attribute.
```
possibly?
```suggestion
`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the requested attribute.
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+
+This proposal suggests to add support for specifying allowed string fields using CEL expressions, similar to the existing
+wildcard support. While CEL expressions could eventually replace the current CSR attribute validation API - including the 
+wildcard support, the main improvement would be support for policies that specifies an allowed attribute value that depends
```
```suggestion
wildcard support, the main improvement would be support for policies that specify an allowed attribute value that depends
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+
+An [experimental approver-policy plugin](https://github.com/erikgb/cel-approver-policy-plugin) providing some of the
+requested features exists, and the proposal (including alternatives) is inspired by this plugin.
+The main downside of using a plugin for this, in addition increased maintenance/complexity,
```
```suggestion
The main downside of using a plugin for this, in addition to increased maintenance/complexity,
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+requested features exists, and the proposal (including alternatives) is inspired by this plugin.
+The main downside of using a plugin for this, in addition increased maintenance/complexity,
+is that the approver-policy rules are still in play.
+Which means that approver rules are somehow duplicated in the `allowed` block and in the plugin configuration and
+should be avoided.
```
```suggestion
Which means that approver rules are somehow duplicated in the `allowed` block and in the plugin configuration but
this should be avoided as maintaining things in multiple places is error prone.
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+information to express the idea and why it was not acceptable.
+-->
+
+### CEL expressions with similar semantics as values
```
```suggestion
### CEL expressions with semantics similar to values
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+
+### CEL expressions with similar semantics as values
+
+_Note: This was the original proposal, after some consideration it is considered inferior to the current proposal,
```
```suggestion
_Note: This was the original proposal, after some consideration it was considered inferior to the current proposal,
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+The idea here is to provide the policy author with a CEL context variable containing the decoded CSR replacing the
+`self` variable in the proposed design. Since the expressions now get access to the full CSR, the expressions will have
+to be "promoted" in the CertificateRequestPolicy API. The expression will have to return a bool indicating approved or
+denied. An example policy with this alternative API might look like (example);
```
```suggestion
denied. An example policy with this alternative API might look like (example):
```

inteon commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+    dnsNames:
+      expressions:
+        - "self.endsWith(cr.namespace + '.svc')"
+        - "self.endsWith(cr.namespace + '.svc.cluster.local')"
```
Maybe, it would be useful if we could also express that eg. you can have at most 3 dnsNames in your cert?

inteon commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+    dnsNames:
+      validations:
+        - message: dnsName can be maximum 64 characters.
+          rule: self.size() =< 64
```
How could we limit the number of dnsNames? Could we eg. have a "selfs" parameter?

erikgb replied:
This is an interesting and important comment! Strictly speaking, many CSR attributes are multi-valued - so it could make sense to make `self` refer to the list of attribute values in the CSR - which will allow for the validation you are requesting here. But I suspect this will make the feature less user-friendly. The validation example this review comment refers to would then have to be expressed as `self.all(dnsName, dnsName.size() =< 64)` - which is not too bad IMO.

But this could easily lead to user errors like `self.size() =< 64` would now limit the number of dnsNames in the CSR to 63. 😉 Which is probably not what the user had in mind.

erikgb replied:
IMO the most user-friendly is to keep the validations under `allowed`  as per-item validations. This is how the `value`/`values` fields in the current API work. More complex CEL validations, typically cross-cutting concerns could be introduced later outside of the `allowed` API structure. I am updating the proposal with this clarification now.

inteon commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,418 @@
+    dnsNames:
+      validations:
+        - message: only dnsNames in the local namespace service domain is allowed.
+          rule: self.endsWith(cr.namespace + '.svc') || self.endsWith(cr.namespace + '.svc.cluster.local')
```
I would start with a `cr` value that only contains the basic fields of the CertifiicateRequest resource (name, namespace, annotations, labels, uuid, ....). Once that works, we could also add other properties (like x509 stuff).

erikgb replied:
Agreed, I think we should start even simpler in the initial implementation and provide just CR `namespace` and `name`. I will add this initial MVP to the proposed design.

erikgb replied:
Updated, PTAL!

erikgb commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,369 @@
+
+The last sentence describes the requirement that could need custom CEL function(s) to allow the user to specify allowed
+CSR attributes. We cannot expect the user to include validations for **all** CSR attributes in the expression(s). So I
+think we at least need a function to list allowed CSR attribute usage.
```
I have now looked at the source code, and approver-policy only evaluates a policy against the CSR attributes that approver-policy supports. So you are correct that it might be considered too permissive in this sense. I suggest creating an issue for this if it is something that we should attempt to fix. Changing it would be breaking tho. 😉

SgtCoDFish reviewed (approved):
/lgtm
/approve

Fundamentally, I think we should be experimenting with approver-policy. Now's the time for it!

I'd be much more conservative about this if it were for cert-manager itself... but it's not. I think this is worth a go. It makes sense. It's well written. Thank you for this Erik! 👍

erikgb commented:
I have initiated the work on implementing this new feature: https://github.com/cert-manager/approver-policy/pull/277
Please take a look if you have time.

erikgb commented:
After working on improving the API docs and understanding more of the current API behavior, I have updated the proposed design (and examples with the following):
>Note: To ensure we don't break behaviors in the existing API version, any validations can never allow a request otherwise denied by the `values`/`value` and `required` fields.

I think that is our safest option. The next API version could make this feature even more user-friendly. But I think it will be acceptable as proposed now.

I am almost ready to move over to the implementation of this, so I would prefer a more "formal" approval of this design proposal by merging this PR. 😉 PTAL!

hawksight commented:
Just wanted to add that the design I read just over 2 weeks ago looked great. It covered a few things I had not thought of and there was nothing I disagreed with. Now work has started, I will attempt to look over for understanding.

The only minor thing is that there are a few empty sections in the design doc, might it be worth either a quick comment of `N/A` if the section isn't applicable.

inteon commented:
/retest

erikgb commented:
/unhold
/ok-to-test

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,459 @@
+To configure valid CSR attribute values, the API provides fields for allowed
+`values` (array; any element matching the attribute is sufficient),
+`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the requested attribute.
+Allowed string fields accept wildcards `"*"` within its value(s).
```
```suggestion
Allowed string fields accept wildcards `"*"` within their values.
```
I think...

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,459 @@
+`value` (must match the attribute) or an array of literals/enums (in `usages`) depending on the requested attribute.
+Allowed string fields accept wildcards `"*"` within its value(s).
+
+This proposal suggests to add support for specifying allowed string fields using CEL expressions, similar to the existing
```
```suggestion
This proposal suggests adding support for specifying allowed string fields using CEL expressions, similar to the existing
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,459 @@
+The new fields should contain CEL validations used to validate CSR attribute values with similar
+[behavior](https://cert-manager.io/docs/projects/approver-policy/#allowed) as approver-policy has today,
+but with the following extension: If validations are present for the given CSR attribute, and **ALL**
+validations passes for **ALL* attribute values, the attribute will be allowed.
```
```suggestion
validations pass for **ALL* attribute values, the attribute will be allowed.
```

jsoref commented on design/20230726-cel-policy.md:
```diff
@@ -0,0 +1,459 @@
+
+Note: To ensure we don't break behaviors in the existing API version, the following should be enforced:
+
+* if `required` set to `true` in policy and attribute value empty in request, the request will be denied
+  by policy without running any validations.
+* if `required` set to `true` in policy, the policy will need specify `values`/`value` or `validations`.
```
```suggestion
* if `required` is set to `true` in policy and attribute value empty in request, the request will be denied
  by policy without running any validations.
* if `required` is set to `true` in policy, the policy will need specify `values`/`value` or `validations`.
```

inteon commented:
/lgtm
